### PR TITLE
Handle episode over or task complete

### PR DESCRIPTION
**Pull request by rpartsey**
## Motivation and Context

Handle episode over or task complete: freeze simulation, show appropriate message about task ending/failing.

## How Has This Been Tested
Running Sandbox App locally (MacBook Pro 2021).

## Types of changes

- **Refactoring**
- **Development** 

## Checklist

- [x] My code follows the code style of this project.
- [x] I have updated the documentation if required.
- [x] I have read the [**CONTRIBUTING**](/CONTRIBUTING.md) document.
- [x] I have completed my CLA (see **CONTRIBUTING**)
- [x] I have added tests to cover my changes if required.

### CODE REVIEW

File: examples/siro_sandbox/sandbox_app.py [hunk 1]
```diff
@@ -163,6 +172,13 @@ def on_environment_reset(self):
         ]
         self._goal_positions = [mn.Vector3(pos) for pos in goal_positions_np]
 
+    @property
+    def task_complete(self):
+        config = self.env._config
```
Comment (eundersander) [inline examples/siro_sandbox/sandbox_app.py:177]: can you cache `success_measure_name` on startup? Accessing `config` is apparently super-slow and should generally be avoided per step.

File: habitat-lab/habitat/gui/gui_application.py [hunk 2]
```diff
@@ -150,7 +150,7 @@ def _post_sim_update(self, post_sim_update_dict):
         if "application_cursor" in post_sim_update_dict:
             self.cursor = post_sim_update_dict["application_cursor"]
 
-        if "application_exit" in post_sim_update_dict:
+        if post_sim_update_dict.get("should_exit", False):
```
Comment (eundersander) [inline habitat-lab/habitat/gui/gui_application.py:153]: I prefer `application_exit`. For messages that specifically go to `GuiApplication`, I was trying to make that clear by using the `application` prefix in the name. If you feel strongly, you could rename to `application_should_exit`.

File: examples/siro_sandbox/sandbox_app.py [hunk 3]
```diff
@@ -623,56 +639,65 @@ def get_grasp_release_controls_text():
...
-            controls_str += "W, S: walk\n"
-            controls_str += "Scroll: zoom\n"
-            controls_str += get_grasp_release_controls_text()
-        else:
-            controls_str += "Left-click + drag: rotate camera\n"
-            controls_str += "A, D: turn camera\n"
-            controls_str += "W, S: pan camera\n"
-            controls_str += "O, P: raise or lower camera\n"
-            controls_str += "Scroll: zoom\n"
+        if not (self.env.episode_over or self.task_complete):
```
Comment (eundersander) [inline examples/siro_sandbox/sandbox_app.py:642]: I don't understand the difference between `episode_over` and `task_complete`. Are you getting at whether the task was successful versus failed?

File: examples/siro_sandbox/sandbox_app.py [hunk 4]
```diff
@@ -800,6 +835,10 @@ def depth_to_rgb(obs):
 
         self._update_help_text()
 
+        # store post_sim_update_dict for frozen sim mode
+        # when episode is over or when task is completed
+        self._prev_post_sim_update_dict = post_sim_update_dict
```
Comment (eundersander) [inline examples/siro_sandbox/sandbox_app.py:840]: This feels like a hack. Why is this necessary?

**Review by eundersander — COMMENTED** [review]
Let's discuss my comments and then give me another chance to review before merging.

File: habitat-lab/habitat/gui/gui_application.py (hunk 2 above)
Comment (eundersander) [reply habitat-lab/habitat/gui/gui_application.py:153]: Also how is your version more Pythonic or cleaner (doing a `get` versus `in`)? It seems less Pythonic.

File: examples/siro_sandbox/sandbox_app.py [hunk 5]
```diff
@@ -709,11 +748,10 @@ def sim_update(self, dt):
...
             )
 
-        action = self.ctrl_helper.update(self.obs)
+        action = self.ctrl_helper.update(self._obs)
+        self.env_step(action)
 
-        self.obs = self.env.step(action)
-
-        post_sim_update_dict = {}
+        post_sim_update_dict = {"should_exit": should_exit}
```
Comment (eundersander) [inline examples/siro_sandbox/sandbox_app.py:754]: I don't see a reason to send the message at all if the value here would be False.

File: examples/siro_sandbox/sandbox_app.py (hunk 4 above)
Comment (eundersander) [reply examples/siro_sandbox/sandbox_app.py:840]: At a high level, we should get away from the idea that, because the simulation is paused, we're necessarily going to render the same thing or otherwise "re-use" the `update_dict`. That is not the case. E.g. I can imagine implementing a camera that kinda circles the agent during this time (even though the sim is paused).

Comment (eundersander) [reply examples/siro_sandbox/sandbox_app.py:840]: One thing that probably needs to change is this logic in `replay_gui_app_renderer.py`:
```
        if len(keyframes):
            self._need_render = True
```

When the sim is paused, we may still need to re-render. Again, the example of the camera moving.

File: habitat-lab/habitat/gui/gui_application.py (hunk 2 above)
Comment (rpartsey) [reply habitat-lab/habitat/gui/gui_application.py:153]: > I prefer application_exit. For messages that specifically go to GuiApplication

👍 , I'll revert this key change, makes sense

> Also how is your version more Pythonic or cleaner (doing a get versus in)? It seems less Pythonic.

Earlier `application_exit` was added to post_sim_update_dict only if application exit was initiated (ESC pressed). After that,  exit app condition was evaluated by checking if `"application_exit"` key is present in `post_sim_update_dict`. This was, `application_exit` key's value was never used. In this implementation I check `application_exit` key's value that's why I used `get` method. It seems to me that it is better to have `post_sim_update_dict` with fixed keys instead of variable. For example, if we ever use `application_exit` key then app behaviour should depend on its value not key being present in the dict or not. But I don't feel strongly about this, can undo this change if you wish.

File: examples/siro_sandbox/sandbox_app.py (hunk 3 above)
Comment (rpartsey) [reply examples/siro_sandbox/sandbox_app.py:642]: Yes, episode is over when any of the constraints is violated and task is complete when the `composite_success` measure is True.

File: examples/siro_sandbox/sandbox_app.py (hunk 4 above)
Comment (rpartsey) [reply examples/siro_sandbox/sandbox_app.py]: Oh, got you! I misunderstood the task. Fixed!

File: examples/siro_sandbox/sandbox_app.py [hunk 6]
```diff
@@ -147,19 +147,33 @@ def __init__(self, args, config, gui_input):
...
-    def on_environment_reset(self):
+    def _env_episode_active(self) -> bool:
+        """
+        Returns True if current episode is active:
+        1) not self.env.episode_over - none of the constraints is violated, or
+        2) not self._env_task_complete - success measure value is not True
+        """
+        return not (self.env.episode_over or self._env_task_complete)
+
+    def _compute_action_and_step_env(self):
```
Comment (eundersander) [inline examples/siro_sandbox/sandbox_app.py:172]: Nit: when a function like this maybe or maybe won't do something, I prefer `try_` or `check_` in the name. I.e. `_check_compute_action_and_step_env`.

**Review by eundersander — APPROVED** [review]
I left a comment but otherwise LGTM!
